### PR TITLE
Fix Java 8 optionals in Schema code

### LINKED ISSUE #1249: Java 8 Optional doesn't work
https://github.com/palantir/atlasdb/pull/1162/files purported to add support for Java 8 optional.

In practice, every generated table implements `AtlasDbNamedImmutableTable` which is the interface that has the `getRow` method on it. This method is defined to return a Guava optional, so none of the generated code actually compiles.

There are a few options here:

1. You could revert the change.
**2. You could delete the interface AtlasDbNamedImmutableTable. None of that interface hierarchy is actually used anywhere (I checked internally), so far as I can tell.**

@uschi2000

### PR DESCRIPTION
**Pull request by gsheasby**
- Removed the unused interfaces AtlasDbNamedMutableTable and
AtlasDbNamedImmutableTable.
- Made the ProfileSchema use Java 8 optionals as a proof of concept
that this can work.

Fixes #1249

@ellisjoe for SA.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1501)

### CODE REVIEW

File: docs/source/release_notes/release-notes.rst [hunk 1]
```diff
@@ -43,6 +43,15 @@ develop
          - Change
 
     *    - |fixed|
+         - Fixed schema generation with Java 8 optionals.
+           To use Java8 optionals, supply ``OptionalType.JAVA8`` as an additional constructor argument when creating your ``Schema`` object.
+           (`Pull Request <https://github.com/palantir/atlasdb/pull/1501>`__)
+
+    *    - |devbreak|
+         - The above fix requires all projects to regenerate their schemas, even if not using Java 8 optionals.
```
Comment (jeremyk-91) [inline docs/source/release_notes/release-notes.rst]: Should we mention a part of the mechanics of the fix (in particular, that the `AtlasDbNamedImmutableTable` and `AtlasDbNamedMutableTable` interfaces have been removed)? From ticket #1249 no internal product uses this but others might.

**Comment by gsheasby** [comment]
I have found that a part of our large internal product actually does use this interface. Since it doesn't use either of the methods that leak the Guava Optional, I will simply restore the interface and remove those two methods.

**Comment by jboreiko** [comment]
Seems like that inferface should receive a `@depracated` and maybe a comment explaining that it has been partially pruned.

**Comment by gsheasby** [comment]
@jboreiko so I pushed a [branch](https://github.com/palantir/atlasdb/tree/deprecate/atlasdb-named-immutable-table) up. Not 100% convinced we actually want to deprecate the interface, but open to discussion.

Having some kind of interface might help us to define an API that products can safely use - I guess this was the original intention of the interface, that is now lost in the mists of time?

**Pull request merged**